File: answers/lintcode/Spiral Matrix.cpp
class Solution {
public:
    /**
     * @param matrix a matrix of m x n elements
     * @return an integer array
     */
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        // Write your code here
        vector<int> spiral;
        for (size_t i = 0; i * 2 < matrix.size(); ++ i) {
            for (size_t j = i; j + i < matrix[i].size(); ++ j) {
                spiral.push_back(matrix[i][j]);
            }
            if (i * 2 < matrix[i].size()) {
                for (size_t j = i + 1; j + i < matrix.size() && i < matrix[j].size(); ++ j) {
                    spiral.push_back(matrix[j][matrix[j].size() - 1 - i]);
                }
            }
            if (i * 2 + 1 < matrix.size()) {
                for (size_t j = i + 1; j + i + 1 < matrix[i].size(); ++ j) {
                    spiral.push_back(matrix[matrix.size() - 1 - i][matrix[matrix.size() - 1 - i].size() - 1 - j]);
                }
            }
            if (i * 2 + 1 < matrix[i].size()) {
                for (size_t j = i; j + i + 1 < matrix.size(); ++ j) {
                    spiral.push_back(matrix[matrix.size() - 1 - j][i]);
                }
            }
        }
        return spiral;
    }
};

File: answers/leetcode/Spiral Matrix/Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new LinkedList<>();
        if (matrix == null || matrix.length == 0) {
            return result;
        }

        int rowBegin = 0;
        int rowEnd = matrix.length - 1;
        int columnBegin = 0;
        int columnEnd = matrix[0].length - 1;
        while (rowBegin <= rowEnd && columnBegin <= columnEnd) {
            for (int i = columnBegin; i <= columnEnd; i++) {
                result.add(matrix[rowBegin][i]);
            }
            rowBegin++;
            for (int i = rowBegin; i <= rowEnd; i++) {
                result.add(matrix[i][columnEnd]);
            }
            columnEnd--;
            if (rowBegin <= rowEnd) {
                for (int i = columnEnd; i >= columnBegin; i--) {
                    result.add(matrix[rowEnd][i]);
                }
                rowEnd--;
            }
            if (columnBegin <= columnEnd) {
                for (int i = rowEnd; i >= rowBegin; i--) {
                    result.add(matrix[i][columnBegin]);
                }
                columnBegin++;
            }
        }
        return result;
    }
}